### 1. Scenario: A company wants to analyze the sales performance of its products in different regions. They have collected the following data:
    Region A: [10, 15, 12, 8, 14]
    Region B: [18, 20, 16, 22, 25]
### calculate the mean sales for each region. 


In [5]:
def cal_mean(sales):
    total_sales = sum(sales)
    mean = total_sales / len(sales)
    return mean
    
sales_Region_A = [10, 15, 12, 8, 14]
mean_A = cal_mean(sales_Region_A)
print(f"Mean Sales for Region A :{mean_A}")

sales_Region_B = [18, 20, 16, 22, 25]
mean_B = cal_mean(sales_Region_B)
print(f"Mean Sales for Region A :{mean_B}")

Mean Sales for Region A :11.8
Mean Sales for Region A :20.2


### 2. Scenario: A survey is conducted to measure customer satisfaction on a scale of 1 to 5. The data collected is as follows:
       [4, 5, 2, 3, 5, 4, 3, 2, 4, 5]
   ### Calculate the mode of the survey responses.


In [10]:
from collections import Counter
def calculate_mode(responses):
    count = Counter(responses)
    mode = count.most_common(1)[0][0]
    return mode

survey_response =  [4, 5, 2, 3, 5, 4, 3, 2, 4, 5]
mode_response = calculate_mode(survey_response)
print(f"mode of survey response :",mode_response )

mode of survey response : 4


### 3. Scenario: A company wants to compare the salaries of two departments. The salary data for Department A and Department B are as follows:
       Department A: [5000, 6000, 5500, 7000]
       Department B: [4500, 5500, 5800, 6000, 5200]
   ### Calculate the median salary for each department.


In [19]:
import statistics

salary_department_a =  [5000, 6000, 5500, 7000]
median_department_a = statistics.median(salary_department_a)
print("median salary for Department A :",median_department_a)

salary_department_b =  [4500, 5500, 5800, 6000, 5200]
median_department_b = statistics.median(salary_department_b)
print("median salary for Department B :",median_department_b)

median salary for Department A : 5750.0
median salary for Department B : 5500


### 4. Scenario: A data analyst wants to determine the variability in the daily stock prices of a company. The data collected is as follows:
       [25.5, 24.8, 26.1, 25.3, 24.9]
  ### Calculate the range of the stock prices


In [30]:
stock_prc = [25.5, 24.8, 26.1, 25.3, 24.9]
price_range = max(stock_prc) - min(stock_prc)

print("Range of the stock price is :" , price_range)

Range of the stock price is : 1.3000000000000007



### 5. Scenario: A study is conducted to compare the performance of two different teaching methods. The test scores of the students in each group are as follows:
       Group A: [85, 90, 92, 88, 91]
       Group B: [82, 88, 90, 86, 87]
  ### Perform a t-test to determine if there is a significant difference in the mean scores between the two groups.


In [42]:
from scipy import stats

# Test scores for Group A and Group B
group_a_scores = [85, 90, 92, 88, 91]
group_b_scores = [82, 88, 90, 86, 87]

# Perform independent t-test
t_statistic, p_value = stats.ttest_ind(group_a_scores, group_b_scores)

# Print the t-statistic and p-value
print("T-Statistic:", t_statistic)
print("P-Value:", p_value)


T-Statistic: 1.4312528946642733
P-Value: 0.19023970239078333


### 6. Scenario: A company wants to analyze the relationship between advertising expenditure and sales. The data collected is as follows:
       Advertising Expenditure (in thousands): [10, 15, 12, 8, 14]
       Sales (in thousands): [25, 30, 28, 20, 26]
   ### Calculate the correlation coefficient between advertising expenditure and sales.


In [4]:
import numpy as np
advertising_Expenditure = np.array([10, 15, 12, 8, 14])
sales = np.array([25, 30, 28, 20, 26])

#cal correlation coefficient :
correlation_coefficient = np.corrcoef(advertising_Expenditure,sales)[0,1]
print("correlation coefficient between advertising expenditure and sales is : ",correlation_coefficient)

correlation coefficient between advertising expenditure and sales is :  0.8757511375750132


### 7. Scenario: A survey is conducted to measure the heights of a group of people. The data collected is as follows:
       [160, 170, 165, 155, 175, 180, 170]
### Calculate the standard deviation of the heights.


In [5]:
import numpy as np
hights = np.array([160, 170, 165, 155, 175, 180, 170])

#cal std-dev:
standard_deviation = np.std(hights)
print("The Sandard Dviation of Hights is :" ,standard_deviation)

The Sandard Dviation of Hights is : 7.953949089757174



### 8. Scenario: A company wants to analyze the relationship between employee tenure and job satisfaction. The data collected is as follows:
       Employee Tenure (in years): [2, 3, 5, 4, 6, 2, 4]
       Job Satisfaction (on a scale of 1 to 10): [7, 8, 6, 9, 5, 7, 6]
### Perform a linear regression analysis to predict job satisfaction based on employee tenure.


In [32]:
import numpy as np
from sklearn.linear_model import LinearRegression

employee_tenure = np.array([2, 3, 5, 4, 6, 2, 4]).reshape(-1,1)   #Scikit-learns LinearRegression excepts 2D input
job_satisfaction = np.array([7, 8, 6, 9, 5, 7, 6])

#create the LinearRegression model and fit the data..
model = LinearRegression()
model.fit(employee_tenure,job_satisfaction)

#Retrieve the slope (coefficient) and intercept of the linear regression line..
coef = model.coef_[0]
intercept = model.intercept_
print(f"""slop of coef is :{coef} and  
intercept is : {intercept}""")

#Take the New employee tenure values for prediction into employee_tenure ...
new_employee_tenure = np.array([2,5,6]).reshape(-1,1)

#prediction job satisfaction for new employee tenure values...
predication_job_satisfaction = model.predict(new_employee_tenure)
print("prediction job satisfaction for new employee tenure : ", predication_job_satisfaction)

slop of coef is :-0.4680851063829787 and  
intercept is : 8.595744680851062
prediction job satisfaction for new employee tenure :  [7.65957447 6.25531915 5.78723404]


### 9. Scenario: A study is conducted to compare the effectiveness of two different medications. The recovery times of the patients in each group are as follows:
       Medication A: [10, 12, 14, 11, 13]
       Medication B: [15, 17, 16, 14, 18]
   ### Perform an analysis of variance (ANOVA) to determine if there is a significant difference in the mean recovery times between the two medications.


In [45]:
from scipy import stats

#recovery times of patients for medications A and B..
medication_a = [10, 12, 14, 11, 13]
medication_b = [15, 17, 16, 14, 18]
#ANOVA:
f_value, p_value = stats.f_oneway(medication_a,medication_b)

print("f_value :",f_value) #measure the difference between the group means.
print("p_value:",p_value) #p_value indocates significance of the observed difference.

f_value : 16.0
p_value: 0.003949772803445326


### 10. Scenario: A company wants to analyze customer feedback ratings on a scale of 1 to 10. The data collected is as follows:
        [8, 9, 7, 6, 8, 10, 9, 8, 7, 8]
### Calculate the 75th percentile of the feedback ratings.



In [46]:
import numpy as np
#Feedback rating:
feedback_ratings = [8, 9, 7, 6, 8, 10, 9, 8, 7, 8]

#cal the 75th percentile...
percentile_75th = np.percentile(feedback_ratings, 75)
print(" 75th percentile of the feedback ratings is :",percentile_75th)

 75th percentile of the feedback ratings is : 8.75


### 11. Scenario: A quality control department wants to test the weight consistency of a product. The weights of a sample of products are as follows:
    [10.2, 9.8, 10.0, 10.5, 10.3, 10.1]
### Perform a hypothesis test to determine if the mean weight differs significantly from 10 grams.


In [47]:
from scipy import stats

weights = [10.2, 9.8, 10.0, 10.5, 10.3, 10.1]

#one-sample t-test to determine if the mean weight differs significantly from 10 grams..
t_test , p_value = stats.ttest_1samp(weights,10)

print("t_test:", t_test)
print("p_value :",p_value)

t_test: 1.5126584522688367
p_value : 0.19077595151110102



### 12. Scenario: A company wants to analyze the click-through rates of two different website designs. The number of clicks for each design is as follows:
    Design A: [100, 120, 110, 90, 95]
    Design B: [80, 85, 90, 95, 100]
### Perform a chi-square test to determine if there is a significant difference in the click-through rates between the two designs.


In [49]:
from scipy import stats

design_a = [100, 120, 110, 90, 95]
design_b = [80, 85, 90, 95, 100]

# Cal observed freq
observed_frequencies = [len(design_a), len(design_b)]

# Cal expected freq assuming equal proportions
total_count = len(design_a) + len(design_b)
expected_frequency = total_count / 2
expected_frequencies = [expected_frequency, expected_frequency]

# Perform chi-square test
chi2_statistic, p_value = stats.chisquare(observed_frequencies, expected_frequencies)

# Print the results
print("Chi-square statistic:", chi2_statistic)
print("p-value:", p_value)


Chi-square statistic: 0.0
p-value: 1.0



### 13. Scenario: A survey is conducted to measure customer satisfaction with a product on a scale of 1 to 10. The data collected is as follows:
    [7, 9, 6, 8, 10, 7, 8, 9, 7, 8]
### Calculate the 95% confidence interval for the population mean satisfaction score.


In [62]:
import math
data = [7, 9, 6, 8, 10, 7, 8, 9, 7, 8]
# calculate sample of mean:
sample_mean = sum(data)/ len(data)

#cal standard deviation:
sample_std_dev = math.sqrt(sum((x - sample_mean) ** 2 for x in data) / (len(data) - 1))
#set the confidence level:
confidence_level = 0.95

#cal z-score 
z_score = 1.96    #in this case, 95% confidence corresponds to a z-score of approximately 1.96

#cal standard error:
standard_error = sample_std_dev / math.sqrt(len(data))

# Cal the margin of error
margin_of_error = z_score * standard_error

# Cal the lower and upper bounds of the confidence interval
lower_bound = sample_mean - margin_of_error
upper_bound = sample_mean + margin_of_error

# Print the confidence interval
print(f"The 95% confidence interval for the population mean satisfaction score is: ({lower_bound:.3f}, {upper_bound:.3f})")

The 95% confidence interval for the population mean satisfaction score is: (7.158, 8.642)


### 14. Scenario: A company wants to analyze the effect of temperature on product performance. The data collected is as follows:
    Temperature (in degrees Celsius): [20, 22, 23, 19, 21]
    Performance (on a scale of 1 to 10): [8, 7, 9, 6, 8]
### Perform a simple linear regression to predict performance based on temperature.


In [68]:
import numpy as np
from sklearn.linear_model import LinearRegression

# Temperature data
temperature = np.array([20, 22, 23, 19, 21])

# Performance data
performance = np.array([8, 7, 9, 6, 8])

# Reshape the temperature data
X = temperature.reshape((-1, 1))

# Reshape the performance data
y = performance

# Create a linear regression model
model = LinearRegression()

# Fit the model to the data
model.fit(X, y)

# Extract the slope (coefficient) of the linear regression line
slope = model.coef_[0]

# Extract the intercept of the linear regression line
intercept = model.intercept_

# Print the equation of the regression line
print(f"Performance = {intercept:.3f} + {slope:.3f} * Temperature")


Performance = -2.900 + 0.500 * Temperature


the performance is expected to increase by 0.500 units.

### 15. Scenario: A study is conducted to compare the preferences of two groups of participants. The preferences are measured on a Likert scale from 1 to 5. The data collected is as follows:
    Group A: [4, 3, 5, 2, 4]
    Group B: [3, 2, 4, 3, 3]
### Perform a Mann-Whitney U test to determine if there is a significant difference in the median preferences between the two groups.


In [73]:
from scipy import stats
group_a = [4, 3, 5, 2, 4]
group_b = [3, 2, 4, 3, 3]

#perform u_test:
u_test , p_value = stats.mannwhitneyu(group_a,group_b)

print("Mann-Whitney U test :",u_test) #difference between the ranked preferences of the two groups
print("p_value :",p_value) #indicates the significance of the observed difference

Mann-Whitney U test : 17.0
p_value : 0.380836480306712


### 16. Scenario: A company wants to analyze the distribution of customer ages. The data collected is as follows:
    [25, 30, 35, 40, 45, 50, 55, 60, 65, 70]
### Calculate the interquartile range (IQR) of the ages.


In [81]:
import numpy as np

# Customer ages data
ages = np.array([25, 30, 35, 40, 45, 50, 55, 60, 65, 70])

# Calculate the 25th and 75th percentiles
q1 = np.percentile(ages, 25)
q3 = np.percentile(ages, 75)

# Calculate the interquartile range (IQR)
iqr = q3 - q1

print("Interquartile Range (IQR):", iqr)


Interquartile Range (IQR): 22.5


indicating that the middle 50% of the ages fall within the range of 22.5 years.

### 18. Scenario: A company wants to analyze the effect of price on sales. The data collected is as follows:
    Price (in dollars): [10, 15, 12, 8, 14]
    Sales: [100, 80, 90, 110, 95]
### Perform a simple linear regression to predict. sales based on price.



In [86]:
import numpy as np
from sklearn.linear_model import LinearRegression

# Price and sales data
price = np.array([10, 15, 12, 8, 14]).reshape(-1, 1)
sales = np.array([100, 80, 90, 110, 95])

# Create an instance of LinearRegression model and fit the data
model = LinearRegression()
model.fit(price, sales)

# Get the slope and intercept of the regression line
slope = model.coef_[0]
intercept = model.intercept_

# Print the equation of the regression line
print("Sales = {:.2f} * Price + {:.2f}".format(slope, intercept))
print("""This equation indicates that for every dollar increase in price, the sales decrease by $5,
and the intercept represents the estimated sales at a price of $0.""")

Sales = -3.51 * Price + 136.37
This equation indicates that for every dollar increase in price, the sales decrease by $5,
and the intercept represents the estimated sales at a price of $0.



### 19. Scenario: A survey is conducted to measure the satisfaction levels of customers with a new product. The data collected is as follows:
    [7, 8, 9, 6, 8, 7, 9, 7, 8, 7]
### Calculate the standard error of the mean satisfaction score.


In [88]:
import numpy as np

# Satisfaction scores data
satisfaction_scores = np.array([7, 8, 9, 6, 8, 7, 9, 7, 8, 7])

# Calculate the standard deviation
standard_deviation = np.std(satisfaction_scores, ddof=1)

# Calculate the number of observations
num_observations = len(satisfaction_scores)

# Calculate the standard error of the mean
standard_error = standard_deviation / np.sqrt(num_observations)

print("Standard Error of the Mean:", standard_error)


Standard Error of the Mean: 0.30550504633038933


### 20. Scenario: A company wants to analyze the relationship between advertising expenditure and sales. The data collected is as follows:
    Advertising Expenditure (in thousands): [10, 15, 12, 8, 14]
    Sales (in thousands): [25, 30, 28, 20, 26]
### Perform a multiple regression analysis to predict sales based on advertising expenditure.



In [89]:
import numpy as np
import statsmodels.api as sm

# Advertising expenditure and sales data
advertising_expenditure = np.array([10, 15, 12, 8, 14])
sales = np.array([25, 30, 28, 20, 26])

# Add a constant term to the independent variable (advertising expenditure)
X = sm.add_constant(advertising_expenditure)

# Fit the multiple linear regression model
model = sm.OLS(sales, X)
results = model.fit()

# Print the summary of the regression analysis
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.767
Model:                            OLS   Adj. R-squared:                  0.689
Method:                 Least Squares   F-statistic:                     9.872
Date:                Tue, 11 Jul 2023   Prob (F-statistic):             0.0516
Time:                        14:41:20   Log-Likelihood:                -9.5288
No. Observations:                   5   AIC:                             23.06
Df Residuals:                       3   BIC:                             22.28
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.2012      4.429      2.755      0.0

C:\Users\DELL\anaconda3\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "
